### PR TITLE
[df-if II] add additional input checks to ensure the input is divisible by 8

### LINKED ISSUE #7842: deepfloyd stage 2 crashes with tensor size mismatch when image is divisible by 8 (sometimes)
### Describe the bug

DeepFloyd's upstream code supports 8px-aligned inputs for stage II, which I believe the Diffusers implementation is based upon. However, it seems that for certain sizes, there is some unfortunate interaction between the hidden states and the residual hidden states.

I'm not sure if this is something fundamental to the model - if it is, we probably want to understand the conditions under which this problem occurs and provide an error to the user about an incompatible resolution.

### Reproduction

```py
from diffusers import DiffusionPipeline, IFSuperResolutionPipeline
import torch
from PIL import Image
import numpy as np

torch.manual_seed(42)

# Configuration for initial image and desired output
initial_width = 86  # Adjusted width to be one-fourth of 344 (approximately)
initial_height = 64  # Adjusted height to be one-fourth of 256

# Initialize your device setting based on availability
torch_device = "cuda" if torch.cuda.is_available() else "mps" if torch.backends.mps.is_available() else "xpu" if torch.xpu.is_available() else "cpu"

# Create a dummy image (86x64)
dummy_image = torch.rand((3, initial_height, initial_width), dtype=torch.float32)  # Random noise image
dummy_image = (dummy_image * 255).to(torch.uint8)  # Convert to 8-bit format
dummy_pil_image = Image.fromarray(dummy_image.numpy().transpose(1, 2, 0))  # Convert to PIL image for compatibility
dummy_pil_image.save("dummy_input.png")  # Save the initial dummy image

# Load your stage 2 pipeline
stage2_pipe = IFSuperResolutionPipeline.from_pretrained("DeepFloyd/IF-II-M-v1.0", watermarker=None, safety_checker=None, local_files_only=False).to(device=torch_device, dtype=torch.bfloat16)

# Upscale the dummy image using stage 2 of the pipeline
upscaled_image = stage2_pipe(
    prompt="A simple upscaled image", 
    image=dummy_pil_image, 
    guidance_scale=5.5, 
    num_inference_steps=20, 
    width=344, 
    height=256
).images[0]

upscaled_image.save("upscaled_dummy_output.png")
```

### Logs

```shell
0%|                                                                                                                                                                                                                                                                                             | 0/20 [00:00<?, ?it/s]

hidden_states.shape: torch.Size([2, 768, 16, 21])
res_hidden_states.shape: torch.Size([2, 768, 16, 21])
hidden_states.shape: torch.Size([2, 768, 16, 21])
res_hidden_states.shape: torch.Size([2, 768, 16, 21])
hidden_states.shape: torch.Size([2, 768, 16, 21])
res_hidden_states.shape: torch.Size([2, 768, 16, 21])
hidden_states.shape: torch.Size([2, 768, 32, 42])
res_hidden_states.shape: torch.Size([2, 768, 32, 43])
```
```


### System Info

- `diffusers` version: 0.27.2
- Platform: macOS-14.4.1-arm64-arm-64bit
- Python version: 3.10.14
- PyTorch version (GPU?): 2.4.0.dev20240421 (False)
- Huggingface_hub version: 0.22.2
- Transformers version: 4.40.0.dev0
- Accelerate version: 0.26.1

### Who can help?

@DN6 @yiyixuxu

### PR DESCRIPTION
**Pull request by bghira**
# What does this PR do?

Fixes #7842 

Adds logic to check_inputs for the IF SuperResolution pipeline so that the user receives a clear error when attempting to run the pipeline with invalid image sizes for the input.

This is possible to hit when using the super-resolution model for upscaling evaluation images during training, if eg. the target 256 pixel resolution is aligned to 8px intervals and then divided by 4 to obtain the input image size. The stage II **output** resolution will be okay, but the **input** resolution would be wrong.

I suppose there's other ways to hit the problem, but it's always been a bit murky which input is causing the problems.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

- Pipelines:  @sayakpaul @yiyixuxu @DN6

### CODE REVIEW

**Comment by yiyixuxu** [comment]
ohh thanks for looking into this!
What we usually do with our image inputs is that in the preprocessing step we resize it to default height and width that are divisible by `8` https://github.com/huggingface/diffusers/blob/58237364b1780223f48a80256f56408efe7b59a0/src/diffusers/image_processor.py#L407

so I think instead of adding the checks, we should just resize it, we can either adding the resize step to the `preprocess_image` method for the IF pipeline, or we can just refactor the method with the VaeImageProcessor like what we do in rest of the pipelines https://github.com/huggingface/diffusers/blob/58237364b1780223f48a80256f56408efe7b59a0/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_img2img.py#L292

what do you think?

**Comment by bghira** [comment]
i considered it. but because of the nature of this, i didn't really feel comfortable just squishing images on the users' behalf. with the small resolution of the inputs, it really can be noticeably distorted, whereas with SD and SDXL at 512/768/1024px it's far less destructive to adjust the size @yiyixuxu how do you feel about that mindset applied to a 64px model, where it might be somewhere around ~5-7% of the image size we end up adjusting by?

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Comment by bghira** [comment]
@yiyixuxu i notice the quality checks failed because of some unnecessary list comprehension. but when i look at it, it seems like the most reasonable way to do it? is there a better way? i would love to learn 😁

File: src/diffusers/pipelines/deepfloyd_if/pipeline_if_superresolution.py [hunk 1]
```diff
@@ -543,12 +543,27 @@ def check_inputs(
 
         if isinstance(image, list):
             image_batch_size = len(image)
+            # Check that each image is the same size:
+            if not all([i.size == image[0].size for i in image]):
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/deepfloyd_if/pipeline_if_superresolution.py:547]:
```suggestion
            if not all(i.size == image[0].size for i in image):
```

**Comment by yiyixuxu** [comment]
can you run `make style` again?

**Comment by bghira** [comment]
@yiyixuxu done

File: src/diffusers/pipelines/deepfloyd_if/pipeline_if_superresolution.py [hunk 2]
```diff
@@ -543,12 +543,27 @@ def check_inputs(
 
         if isinstance(image, list):
             image_batch_size = len(image)
+            # Check that each image is the same size:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/deepfloyd_if/pipeline_if_superresolution.py:546]: I think it is better to do this in a separate code block:
So we keep this section as it is to check `image_batch_size`

and then
```
if isinstance(image, list):
    check_image_size = image[0]
else:
    check_image_size = image

if isinstance(check_image_size, PIL.Image.Image):
    image_size = check_image_size.size
elif isinstance(check_image_size, torch.Tensor):
   image_size = check_image_size.shape[2:]
elif isinstanc(..., np.ndarray):
    image_size = check_image.shape[:1]

if image_size ....:
    raise ValueError(...)

```

The current code does not work for list of array or tensors

File: tests/pipelines/deepfloyd_if/test_if_superresolution.py [hunk 3]
```diff
@@ -57,6 +57,15 @@ def get_dummy_inputs(self, device, seed=0):
 
         return inputs
 
+    def test_incorrect_input_size(self):
+        # Put an image non-divisible by 8 into the pipeline and check that it throws an Exception.
+        image = floats_tensor((1, 3, 31, 31), rng=random.Random(0)).to(torch_device)
+        generator = torch.Generator(device="cpu").manual_seed(0)
+        with self.assertRaises(ValueError):
+            self.pipeline(
```
Comment (yiyixuxu) [inline tests/pipelines/deepfloyd_if/test_if_superresolution.py:65]: can we make sure this test works?

Comment (bghira) [reply tests/pipelines/deepfloyd_if/test_if_superresolution.py:65]: i can't run the test suite locally, i was waiting for it to run on the workflow here

Comment (yiyixuxu) [reply tests/pipelines/deepfloyd_if/test_if_superresolution.py:65]: oh so these are the tests that failed https://github.com/huggingface/diffusers/actions/runs/9044175339/job/24855064736#step:7:15620
I can trigger them again now but I think the results would be the same
